### ISSUE TITLE
Issues not transitioned to status when using Jira Server

### ISSUE THREAD

**Issue by aufdenpunkt** [Qytera-Gmbh/cypress-xray-plugin#389]
### Description

Hi,

This is actually not a bug in your really nice plugin. We noticed that it is not possible to set the status/transition via the Jira Server REST API, when the issue POST endpoint is used.

```js
...
      await configureXrayPlugin(fixedOn, config, {
        jira: {
          projectKey: 'TTEST',
          url: 'server url',
          testExecutionIssue: {
            transition: {
              id: '6',
            },
          },
        },
      });
...
```

Unfortunately, when it comes to transitions, the Jira Server and Jira Cloud REST APIs work slightly differently. See https://jira.atlassian.com/browse/JRASERVER-66881

Is it possible to implement some kind of "hooks" that can be used to make another request to the API per issue, e.g. to call an additional endpoint to set the status?

Thanks and best,
Chris

### Cypress version

13.14.2

### Plugin version

7.3.0

### Jira/Xray type

Server

### Acknowledgements

- [X] I have made sure not to include any sensitive data in screenshots or obfuscated all occurrences.
- [X] I have made sure not to include any sensitive data in descriptions or replaced all values with placeholders.
- [X] I have made sure not to include any sensitive data in code snippets or replaced all values with placeholders.

**Comment by csvtuda** [comment]
Hi @aufdenpunkt,

interesting, I've never actually used the `transition` property.

> Is it possible to implement some kind of "hooks" that can be used to make another request to the API per issue, e.g. to call an additional endpoint to set the status?

This would work, yes. Maybe something like:

```ts
await configureXrayPlugin(fixedOn, config, {
  jira: {
    projectKey: 'TTEST',
    url: 'server url',
    // https://docs.atlassian.com/software/jira/docs/api/REST/7.6.1/#api/2/issue-createIssue
    onResultUpload: (issue, jiraClient, xrayClient) => {
      // call transition issue endpoint yourself here
    }
  },
});
```

I could optionally make the clients my plugin uses available there, too. Some of the endpoints there could maybe be reused.

---

Alternatively, I can also implement a `transition` detection and perform transition calls in the plugin, too. Shouldn't be too difficult.

Let me know what you think.

**Comment by aufdenpunkt** [comment]
Hi @csvtuda,

thanks for your quick response. So, basically, having hooks would enable a certain flexibility. But, as this is more an edge case in my opinion, I would prefer the transition detecktion.

**Comment by csvtuda** [comment]
A quick update: I have not been able to work on this yet. I'll probably implement it this weekend, there's a lot of other stuff going on at the moment.

**Comment by aufdenpunkt** [comment]
Don't worry and thanks for the update. Let me know if I can support you in any way or can contribute.
Have a great weekend.

**Comment by csvtuda** [comment]
Hi @aufdenpunkt,

I have released version 7.4.0, which now explicitly performs the transition iff:

- a transition is specified in `jira.testExecutionIssue.transition`
- no existing test execution issue is reused (`jira.testExecutionIssue.key` / `jira.testExecutionIssueKey`)
- Jira server is used

The code from your example should hopefully work now. Let me know if it does!

**Comment by aufdenpunkt** [comment]
Hi @csvtuda,

Sorry for my late response, and thanks for the release! I will have a look in the upcoming days.

**Comment by aufdenpunkt** [comment]
Hi @csvtuda,

It is working now, thank you very much. 🙂

During my testing, I stepped into a minor bug. Setting the transition ID as a property inside the transition object, like in my example configuration above, leads to a `400 Bad request`. After some debugging, I've figured out that the request has been made with the body `{ "id": "123" }`. But the body needs to be like `{ "transition": { "id": "123" } }`. I was able to fix it using a workaround, wrapping it into another `transition` object like so:

```
...
      await configureXrayPlugin(fixedOn, config, {
        jira: {
          projectKey: 'TTEST',
          url: 'server url',
          testExecutionIssue: {
            transition: {
              transition: {
                id: '6',
              },
            },
          },
        },
      });
...
```
